### PR TITLE
Prevent to save too large module content > 65535

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #24631 .

### Summary of Changes
added a check to prevent to save too large module content > 65535

### Testing Instructions

- Create a module containing a field type of editor.
- Using this module, enter any content which exceeds 65535 characters.

### Expected result
an error message is displayed that you are not permitted to save too large content 

### Actual result
no check

### CODE REVIEW

File: administrator/components/com_modules/models/module.php [hunk 1]
```diff
@@ -907,6 +907,14 @@ public function save($data)
 		$isNew      = true;
 		$context    = $this->option . '.' . $this->name;
 
+		// Prevent to save too large content > 65535 
+		if (strlen($data['content']) > 65535)
```
Comment (mbabker) [inline administrator/components/com_modules/models/module.php]: This type of validation belongs in the table class, not the model.

If character limitations are going to be checked, then this should be applied consistently across core and not only on this field IMO.  Then again, from what I remember at least MySQL throws warnings about right truncating columns when the values are too big.  Is this not happening consistently in Joomla, and if so why not!?  It seems like that should be an Exception that someone can catch and handle more gracefully than this type of check.

**Comment by ghost** [comment]
@ricomonkeon please test.

**Comment by ghost** [comment]
@alikon is it okay for you if i copy the testing instructions from #24631?

File: administrator/components/com_modules/models/module.php (hunk 1 above)
Comment (alikon) [reply administrator/components/com_modules/models/module.php]: you are right as always, btw, should be a little bit better now....not perfect but....
check moved on table class

**Comment by alikon** [comment]
@franz-wohlkoenig  feel free to do it

File: libraries/src/Table/Module.php [hunk 2]
```diff
@@ -123,6 +123,14 @@ public function check()
 			return false;
 		}
 
+		// Prevent to save too large content > 65535 
+		if (strlen($this->content) > 65535)
+		{
+			$this->setError(\JTEXT::_('COM_MODULES_FIELD_CONTENT_TOO_LARGE'));
+
```
Comment (Quy) [inline libraries/src/Table/Module.php]: Remove tabs.

File: libraries/src/Table/Module.php [hunk 3]
```diff
@@ -123,6 +123,14 @@ public function check()
 			return false;
 		}
 
+		// Prevent to save too large content > 65535 
+		if (strlen($this->content) > 65535)
+		{
+			$this->setError(\JTEXT::_('COM_MODULES_FIELD_CONTENT_TOO_LARGE'));
```
Comment (Quy) [inline libraries/src/Table/Module.php]:
```suggestion
			$this->setError(\JText::_('COM_MODULES_FIELD_CONTENT_TOO_LARGE'));
```

**Comment by ricomonkeon** [comment]
Just tested it.  I'm out of my depth in this part of Joomla, but should it not be $data[content] instead of $this->content ? (the $this object does not seem to contain the submitted data)
Also, note that if you add an editor field type in your module, then it does not use the content field, but saves it as part of a json string in the params field with the rest of the data. Only customContent seems to use the content field.
The failure in $data[content] isn't actually as bad - it's the broken json string in params which causes a fatal error.

**Comment by alikon** [comment]
well i've only experienced  with `content` .... btw added the check on `params` too 
please re-test

**Comment by ricomonkeon** [comment]
$this->content and $this->params always seem to return a length of zero.

It feels like -
if ((strlen($this->content) > 65535) || (strlen($this->params) > 65535))
Should be something like -
if ((strlen($data[content]) > 65535) || (strlen(json_encode($data[params])) > 65535))
The latter works on my test, but I'm not sure how tidy my code is.

**Comment by alikon** [comment]
can you post the zip of your module please

**Comment by mbabker** [comment]
> t feels like -
> if ((strlen($this->content) > 65535) || (strlen($this->params) > 65535))
> Should be something like -
> if ((strlen($data[content]) > 65535) || (strlen(json_encode($data[params])) > 65535))

Make sure you're editing the right file.  In the table's `check()` method, the properties are bound to the class before it is called so `$this->foo` should be what you're using in that context.

If you're having to check `$data['foo']` then you might be editing the wrong file.

**Comment by ricomonkeon** [comment]
Apologies.  I was indeed editing the wrong file.  
The bug is now fixed - good work!

**Comment by ghost** [comment]
@ricomonkeon please mark your test in Issue Tracker as successfully > https://docs.joomla.org/Testing_Joomla!_patches#Recording_test_results

**Comment by HLeithner** [comment]
2 questions.

1. Why do you limit the params mediumtext column to 65k chars? it should be 16m
2. Why do we limit to text, we as joomla has no gain from this, especially if we don't have another column that can be used for meta data.

params would be such column but it seams that it isn't used by devs for there needs, why don't they use it and would a new column dedicated for easy use better?

**Comment by HLeithner** [comment]
and btw. this is a real bad user experience if you edit something and the system tells you I can't save it without any help how to solve it. Especially when extensions are abusing this field and fill it with json data that isn't seen or can't be modified by the user.

**Comment by alikon** [comment]
both field `content` and `params` are declared as `text`
https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L1498
https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L1509

so  65,535 characters should be the  max  length allowed iirc

the goal of this pr is really simple: only to prevent data truncation, it's only a band-aid 
not a schema redesign nor a complete review of Exception flows as correctly Michael pointed out https://github.com/joomla/joomla-cms/pull/24635#discussion_r276759676

**Comment by HLeithner** [comment]
sorry about the mediumtext, the first website I checked had it as mediumtext and I don't know why.

ok then I understand thx.

**Comment by ricomonkeon** [comment]
I have tested this item :white_check_mark: successfully on 406da0131633f157afae62fb284677d9423b653e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24635">issues.joomla.org/tracker/joomla-cms/24635</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 93e3505b49daef3d3379408b5d77798a35e57090<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24635">issues.joomla.org/tracker/joomla-cms/24635</a>.</sub>

**Comment by ghost** [comment]
@HLeithner set this on RTC or retest of @ricomonkeon needed?

**Comment by zero-24** [comment]
No need to retest. We just need to get appveyor running. Let me reboot him

**Comment by zero-24** [comment]
https://ci.appveyor.com/project/release-joomla/joomla-cms/builds/23987426

**Comment by ghost** [comment]
Status "Ready To Commit".

**Comment by anirudh7** [comment]
But why text limitations? I have 60+ Speakers attending the conference, and I want to use as a module in multiple paces.

**Pull request merged**